sdk/src/main/java/com/bugsnag/android/FileStore.java

```diff
@@ -86,22 +89,50 @@ String write(@NonNull T streamable) {
         return null;
     }
 
-    @NonNull abstract String getFilename(T streamable);
+    @NonNull
+    abstract String getFilename(T streamable);
 
     List<File> findStoredFiles() {
         List<File> files = new ArrayList<>();
 
-        if (storeDirectory != null) {
-            File dir = new File(storeDirectory);
+        if (oldDirectory != null) {
+            File dir = new File(oldDirectory);
+            addStoredFiles(dir, files);
+        }
+        if (storageDir != null) {
+            addStoredFiles(storageDir, files);
+        }
+        return files;
+    }
 
-            if (dir.exists() && dir.isDirectory()) {
-                File[] values = dir.listFiles();
+    void addStoredFiles(File dir, List<File> files) {
+        if (!dir.exists() || !dir.isDirectory()) {
+            return;
+        }
+        File[] values = dir.listFiles();
 
-                if (values != null) {
-                    files.addAll(Arrays.asList(values));
+        if (values != null) {
+            for (File value : values) {
+                if (value.isFile()) {
+                    files.add(value);
                 }
             }
         }
-        return files;
     }
+
+    // support multiple clients in the same app by using a unique directory path
+
+    private File getStorageDir(String path, @NonNull Configuration config) {
```

Comment:
If a project rotates API keys or the on-premise endpoint changes, will stored files ever be sent (or cleared)?

Reply:
Not currently. Is this something we need to address?

Reply:
Yeah - in the worst case, files shouldn't be left on disk indefinitely. In the best case, everything is delivered according to the configuration at the time the report was captured. Splitting directories by endpoint and key works until  either component needs to change. A better approach would be to write the entire request to disk including the endpoint and the headers. For example, a (nearly) raw HTTP request (format amended to encode protocol and port):

```http
POST https://notify.bugsnag.com
Bugsnag-Api-Key: aaaabaaaabaaaabaaaa
Content-Type: application/json

{"events": [{"severity": "warning", "exceptions":[]}]
```

on premise example for contrast:

```http
POST https://bugsnag.example.internal:79002
Bugsnag-Api-Key: aaaabaaaabaaaabaaaa
Content-Type: application/json

{"events": [{"severity": "warning", "exceptions":[]}]
```

Which could then be parsed until encountering the double newline for the endpoint and headers, then stream the rest as the body directly.

This should:

* Avoid cycling directories, ensuring everything is delivered and cleared from disk
* Tie API key to a request even if cached on disk
* Not affect reports not cached to disk
* Have the same integration tests
* Have the same/similar logic for handling reports in the old location

Reply:
I think this approach makes sense.

Reply:
https://developer.android.com/reference/android/content/Context#getcachedir